discovery/sync_manager.go

```diff
@@ -494,6 +543,94 @@ func (m *SyncManager) isPinnedSyncer(s *GossipSyncer) bool {
 	return isPinnedSyncer
 }
 
+// maybeAddMsgDelay will add a delay to the gossip syncer if the peer has been
+// requesting too much.
+func (m *SyncManager) maybeAddMsgDelay(ctx context.Context,
+	peerPub [33]byte, msg lnwire.Message) (time.Duration, error) {
+
+	var (
+		msgSize uint32
+		err     error
+	)
+
+	// Figure out the serialized size of the message. If we can't easily
+	// compute it, then we'll used the assumed msg size.
+	if sMsg, ok := msg.(lnwire.SizeableMessage); ok {
+		msgSize, err = sMsg.SerializedSize()
+		if err != nil {
+			return 0, err
+		}
+	} else {
+		log.Warnf("Unable to compute serialized size of %T", msg)
+
+		msgSize = assumedMsgSize
+	}
+
+	reservation := m.rateLimiter.ReserveN(time.Now(), int(msgSize))
+	delay := reservation.Delay()
+
+	return delay, nil
+}
+
+// waitMsgDelay takes a delay, and waits until it has finished.
+func (m *SyncManager) waitMsgDelay(ctx context.Context, peerPub [33]byte,
+	delay time.Duration) error {
+
+	// If we've already replied a handful of times, we will start to delay
+	// responses back to the remote peer. This can help prevent DOS attacks
+	// where the remote peer spams us endlessly.
+	if delay > 0 {
+		log.Infof("GossipSyncer(%x): rate limiting gossip replies, "+
+			"responding in %s", peerPub, delay)
+
+		select {
+		case <-time.After(delay):
+
+		case <-ctx.Done():
+			return ErrGossipSyncerExiting
+
+		case <-m.quit:
+			return ErrGossipSyncerExiting
+		}
+	}
+
+	return nil
+}
+
+// maybeRateLimitMsg takes a message, and may wait a period of time to rate
+// limit the msg.
+func (m *SyncManager) maybeRateLimitMsg(ctx context.Context, peerPub [33]byte,
+	msg lnwire.Message) error {
+
+	delay, err := m.maybeAddMsgDelay(ctx, peerPub, msg)
```

Comment:
It looks like we can just call `m.rateLimiter.Wait` here, which combines the two methods `maybeAddMsgDelay` and `waitMsgDelay`.

Reply:
So `Wait` will only attempt to consume a single token, a token here is a single byte. `WaitN` is similar to `ReserveN`, but it'll actually _error out_ if the rate limit is exceeded. Rather than error out and drop _our_ send, I think we want to just wait until more room is available in the bucket.

Calling `ReserveN` instead is nice as it lets us log the delay (it returns the delay for you to sleep yourself instead of the library sleeping for you). 

I split it up into two methods so we can log the delay, and we can also use `maybeAddDelayMsg` for unit tests, as it returns the delay instead of sleeping itself.